zio-http/shared/src/main/scala/zio/http/codec/HttpContentCodec.scala

```diff
@@ -0,0 +1,195 @@
+package zio.http.codec
+
+import scala.collection.immutable.ListMap
+
+import zio._
+
+import zio.stream.ZPipeline
+
+import zio.schema.Schema
+import zio.schema.codec._
+
+import zio.http._
+import zio.http.codec.HttpContentCodec.json.fromSchema
+
+final case class HttpContentCodec[A](
+  choices: ListMap[MediaType, BinaryCodec[A]],
+  schema: Schema[A],
+) { self =>
+
+  /**
+   * A right biased merge of two HttpContentCodecs.
+   */
+  def ++(that: HttpContentCodec[A]): HttpContentCodec[A] =
+    HttpContentCodec(choices ++ that.choices, schema)
+
+  def decodeRequest(request: Request): Task[A]    = {
+    val contentType =
```

Comment:
All this code is in the hot path, which means that it should be heavily optimized to avoid allocations.

One example: lookup the codec using `contains` and `get` instead of allocating an `Option` with `getOrElse`. Also, do not continuously create a new `Header.ContentType`, but rather, define `Header.ContentType.application/json` so that you can reuse the same one again and again.

Reply:
I know that there needs to be optimizations. I made it work first. And wanted some feedback for the implications

Reply:
Actually, the encode and decode methods are not used at all. The magic is happening in `BodyCodec`